storage/engine.go

```diff
@@ -662,6 +667,76 @@ func (e *Engine) deleteBucketRangeLocked(ctx context.Context, orgID, bucketID pl
 	return e.engine.DeletePrefixRange(ctx, name, min, max, pred)
 }
 
+func (e *Engine) CreateBackup(ctx context.Context) (int, []string, error) {
+	span, ctx := tracing.StartSpanFromContext(ctx)
+	span.LogKV("path", e.path)
+
+	err := e.engine.WriteSnapshot(ctx, tsm1.CacheStatusBackup)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	id, snapshotPath, err := e.engine.FileStore.CreateSnapshot(ctx)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	fileInfos, err := ioutil.ReadDir(snapshotPath)
+	if err != nil {
+		return 0, nil, err
+	}
+	filenames := make([]string, len(fileInfos))
+	for i, fi := range fileInfos {
+		filenames[i] = fi.Name()
+	}
+
+	return id, filenames, nil
+}
+
+func (e *Engine) FetchBackupFile(ctx context.Context, backupID int, backupFile string, w io.Writer) error {
+	span, ctx := tracing.StartSpanFromContext(ctx)
+	span.LogKV("path", e.path)
+
+	backupPath := e.engine.FileStore.InternalBackupPath(backupID)
+	if fi, err := os.Stat(backupPath); err != nil {
+		if os.IsNotExist(err) {
+			return errors.Errorf("backup %d not found", backupID)
+		}
+		return errors.WithMessagef(err, "failed to locate backup %d", backupID)
+	} else {
+		if !fi.IsDir() {
+			return errors.Errorf("error in filesystem path of backup %d", backupID)
+		}
+	}
+
+	backupFileFullPath := filepath.Join(backupPath, backupFile)
+	file, err := os.Open(backupFileFullPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return errors.Errorf("backup file %d/%s not found", backupID, backupFile)
+		}
+		return errors.WithMessagef(err, "failed to open backup file %d/%s", backupID, backupFile)
+	}
+	defer file.Close()
+
+	buf := make([]byte, 1024*1024)
+	_, err = io.CopyBuffer(w, file, buf)
+	if err != nil {
+		return errors.WithMessagef(err, "failed to copy backup file %d/%s to writer", backupID, backupFile)
+	}
+
+	err = syscall.Unlink(backupFileFullPath)
```

Comment:
I think this is the first time I have seen `syscall.Unlink` in the Influx codebase. At least outside of the context of a socket. I expect this is because the file is linked when being created in the path, but I don't recall previous backup/restore APIs in influx unlinking, so i'm interested :-)

Reply:
`os.Remove` removes either a file or a directory. I know this is a file, so I shortcut to `syscall.Unlink`. I just confirmed that we don't use `syscall.Unlink` anywhere, and I don't care to defend this, so `os.Remove` it is.

For the curious:
```golang
package os

func Remove(name string) error {
	// System call interface forces us to know
	// whether name is a file or directory.
	// Try both: it is cheaper on average than
	// doing a Stat plus the right one.
	e := syscall.Unlink(name)
	if e == nil {
		return nil
	}
	e1 := syscall.Rmdir(name)
	if e1 == nil {
		return nil
	}

	// Both failed: figure out which error to return.
	// OS X and Linux differ on whether unlink(dir)
	// returns EISDIR, so can't use that. However,
	// both agree that rmdir(file) returns ENOTDIR,
	// so we can use that to decide which error is real.
	// Rmdir might also return ENOTDIR if given a bad
	// file path, like /etc/passwd/foo, but in that case,
	// both errors will be ENOTDIR, so it's okay to
	// use the error from unlink.
	if e1 != syscall.ENOTDIR {
		e = e1
	}
	return &PathError{"remove", name, e}
}
```